nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -205,120 +212,54 @@ Costmap2DROS::~Costmap2DROS()
-    map["type"] = XmlRpc::XmlRpcValue("nav2_costmap_2d::ObstacleLayer");
-    super_map.setStruct(&map);
-    plugins.push_back(super_map);
-  }
-
-  move_parameter(nh, obstacles, "max_obstacle_height");
-  move_parameter(nh, obstacles, "raytrace_range");
-  move_parameter(nh, obstacles, "obstacle_range");
-  move_parameter(nh, obstacles, "track_unknown_space", true);
-
-  s = parameters_client->get_parameter<std::string>("observation_sources", std::string(""));
-  std::stringstream ss(s);
-  std::string source;
-  while (ss >> source) {
-    move_parameter(nh, obstacles, source);
-  }
-  move_parameter(nh, obstacles, "observation_sources");
+  std::vector<rclcpp::Parameter> param;
 
-  auto inflation = rclcpp::Node::make_shared("obstacle_layer", std::string(nh->get_name()));
-  //ros::NodeHandle inflation(nh, "inflation_layer");
+  std::vector<std::string> plugin_names = {"static_layer","inflation_layer"};
+  std::vector<std::string> plugin_types = {"nav2_costmap_2d::StaticLayer","nav2_costmap_2d::InflationLayer"};
+  param = {rclcpp::Parameter("plugin_names",plugin_names),rclcpp::Parameter("plugin_types",plugin_types)};
+  node->set_parameters(param);
+}
 
-  move_parameter(nh, inflation, "cost_scaling_factor");
-  move_parameter(nh, inflation, "inflation_radius");
-  map["name"] = XmlRpc::XmlRpcValue("inflation_layer");
-  map["type"] = XmlRpc::XmlRpcValue("nav2_costmap_2d::InflationLayer");
-  super_map.setStruct(&map);
-  plugins.push_back(super_map);
+void Costmap2DROS::reconfigureCB(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  RCLCPP_DEBUG(node_->get_logger(), "Costmap2DROS:: Event Callback");
 
-  super_array.setArray(&plugins);
-  // TODO(bpwilcox): parameter client doesn't support setting xmlrpc data structure
-  //nh.setParam("plugins", super_array);
-  //auto set_parameters_results = parameters_client->set_parameters({
-  //  rclcpp::Parameter("plugins", super_array)
-  //});
+  dynamic_param_client_->get_event_param(event,"transform_tolerance", transform_tolerance_);
 
-}
-// TODO(bpwilcox): resolve dynamic reconfigure dependencies
-/*
-void Costmap2DROS::reconfigureCB(nav2_costmap_2d::Costmap2DConfig &config, uint32_t level)
-{
-  transform_tolerance_ = config.transform_tolerance;
   if (map_update_thread_ != NULL)
   {
     map_update_thread_shutdown_ = true;
     map_update_thread_->join();
     delete map_update_thread_;
   }
   map_update_thread_shutdown_ = false;
-  double map_update_frequency = config.update_frequency;
+  double map_update_frequency = 1.0;
+  dynamic_param_client_->get_event_param(event,"update_frequency", map_update_frequency);
+
+  double map_publish_frequency = 5.0;
+  dynamic_param_client_->get_event_param(event,"publish_frequency", map_publish_frequency);
 
-  double map_publish_frequency = config.publish_frequency;
   if (map_publish_frequency > 0)
-    publish_cycle = ros::Duration(1 / map_publish_frequency);
+    publish_cycle_ = rclcpp::Duration(1*1e9 / map_publish_frequency);
```

Comment:
I don't understand the 1e9,  I believe this would still the seconds constructor

Reply:
If I understand correctly, the input for rclcpp::Duration is in nanoseconds. Since the frequency in the ROS1 case assumed in Hz, (i.e. cycles per second), we need to pass in the time of one cycle for the duration, which in this case should be in nanoseconds, hence the 1e9. Let me know if you believe I missed something.

Reply:
http://docs.ros2.org/ardent/api/rclcpp/classrclcpp_1_1_duration.html

Reply:
In the [duration.cpp](https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/duration.cpp#L40)
`Duration::Duration(int64_t nanoseconds)
{
  rcl_duration_.nanoseconds = nanoseconds;
}`